### ISSUE TITLE
Small black artifacts when map is zoomed in/out

### ISSUE THREAD

**Issue by grz0** [wesnoth/wesnoth#6959]
### Game and System Information

The Battle for Wesnoth version 1.17.6+dev (565b318c9f0-Modified) x86_64
Running on Fedora Linux 36 (Workstation Edition) x86_64
Distribution channel: Default

SDL video drivers:        wayland [x11] KMSDRM dummy
Window size:              1920x1021
Game canvas size:         1920x1021
Final render target size: 1920x1021
Screen refresh rate:      60
Boost:             1.76
Lua:               5.4.4
OpenSSL/libcrypto: 3.0.0e-dev (runtime 3.0.0e-dev)
Cairo:             1.17.6 (runtime 1.17.6)
Pango:             1.50.8 (runtime 1.50.8)
SDL:               2.0.22 (runtime 2.0.22)
SDL_image:         2.0.5 (runtime 2.0.5)
SDL_mixer:         2.0.4 (runtime 2.0.4)

### Description of the bug

This happens only for some zoom levels:
![image](https://user-images.githubusercontent.com/3605541/184409095-82a78d27-847a-42ed-a3b5-0c6286ef0fa8.png)
![image](https://user-images.githubusercontent.com/3605541/184409150-fbb90f22-2db8-49ff-8911-9b59e335873e.png)

### Steps to reproduce the behavior

1. Zoom in/out until you see black artifacts

**Comment by Wedge009** [comment]
The 'artefacts' looks like a slight misalignment of the hex images at certain zoom levels. I observe the issue in current master (314f6f7c78d), but not in 1.16 branch.

**Comment by Wedge009** [comment]
A git bisect suggests this regression started in 46750666b25704b97766bc691a199dee111b56f6, introduced with 1.17.5.

**Comment by gfgtdf** [comment]
Happens to me too, the affected zoom levels are 0.33333333, 0.5 and 1.5

**Comment by gfgtdf** [comment]
M first idea here would be to first assemble the map from it tiles, an then zoom the assembed map instead of zooming the tiles seperately and then placing them on the map, but i have no idea how easy that'd be.

**Comment by Wedge009** [comment]
At an abstract level, likewise without knowing the technical details of the implementation, that does sound like a good approach.

@mesilliac Sorry to bother you, but noticed a few comments from you recently. Is that an accurate representation of what actually happens, that the tiles are being scaled separately?

**Comment by mesilliac** [comment]
Yeah. It was always like this, just the scaling was more predictable when it was all done in software. I didn't get seams at any zoom level when testing so this is probably dependent on GPU.

Rendering at 1:1 then scaling could be onerous if zoomed out a lot. However it should be technically doable as it's already rendering at one res then displaying at another for high-dpi support.

Alternatively it might be possible to get it to scale pixel-perfectly by messing with fractional scaling or offsets of less than a pixel. Like if it's supposed to scale to 24x24 pixels you could tell it to scale to 24.25 x 24.25 and it might remove the seams. Or tell it to place everything half a pixel to the left. Not sure how technically feasible these are but hacks like this are annoyingly common in GPU code as there are a lot of ambiguous edge cases which different GPUs make different decisions about.

**Comment by Wedge009** [comment]
> Not sure how technically feasible these are but hacks like this are annoyingly common in GPU code as there are a lot of ambiguous edge cases which different GPUs make different decisions about.

That's unfortunate. ): For reference, I use primarily AMD GPUs. X11 and Windows.

**Comment by mesilliac** [comment]
I'm on Nvidia. Kind of wish i chose AMD with how much grief i get trying to run Wayland.

**Comment by Wedge009** [comment]
As an aside, that's an interesting anecdote. I constantly hear of how great Nvidia GPUs are on Linux... admittedly only with the proprietary driver. I thought Wayland support was already there? Maybe still WIP, I know there's still plenty of work to go for Wayland generally, but I understand it's 'getting there'.

**Comment by gfgtdf** [comment]
For reference I tested with Intel integrated GPU .

**Comment by Wedge009** [comment]
Perhaps it is indeed GPU/driver-related. Checking again today, it seems to be far less noticeable an issue than when I first verified in August 2022. Currently running Mesa 23.2.1.

Of course, it could also be that my recollection is hazy.

**Comment by mesilliac** [comment]
> I thought Wayland support was already there? Maybe still WIP

It's okay for native Wayland apps. It's just borked for Xwayland compatibility resulting in stuttering, flickering, and frames out of order for X11 apps. Which includes most games and cross-platform apps. Still, it's usable for day-to-day, and SDL apps work fine if you force them to use wayland via environment variable.

It's been a while since i felt like nvidia was great on Linux. Definitely it used to be.

**Comment by gfgtdf** [comment]
> Like if it's supposed to scale to 24x24 pixels you could tell it to scale to 24.25 x 24.25 and it might remove the seams. Or tell it to place everything half a pixel to the left

I'd imagine that scaling by "simple" factor like 0.5 would look better, especially for our pixel art, and that such a change probably makes it look worse, but im not sure.

**Comment by mesilliac** [comment]
The point of the odd value is that it never scales by more than half a pixel - so all pixels get scaled to the exact same integer location after rounding and the full image is still 24 pixels wide. The problem is likely the nearest-neighbour interpolation picking an empty pixel to interpolate from.

For example if you have something like
```
OOOO
OXXO
OXXO
OOOO
```
and you scale it to half size using nearest-neighbour
```
AB
CD
```
it's ambiguous what all of ABCD are, depending on how you define "nearest", as each of them is perfectly in between four pixels.

Scaling or offsetting by less than half a pixel (in the target size) should cause no difference in final rendering, but clears up the ambiguity by making each of ABCD technically a tiny bit nearer to one of the four than the rest.

**Comment by gfgtdf** [comment]
> Scaling or offsetting by less than half a pixel (in the target size) should cause no difference in final rendering, but clears up the ambiguity by making each of ABCD technically a tiny bit nearer to one of the four than the rest.

okay sounds good in theory, but it seems like all api that wee are using only supports integer coordinated, in particular SDL_RenderCopy

**Comment by gfgtdf** [comment]
okay it seems like `SDL_RenderCopyF`is actualyl a thing... maybe we should try that one out eventually.

**Comment by gfgtdf** [comment]
So i tried to fix it by using the mentioned method of making the blit target a little bit larger by 0.25, and i now get gres artifacts on fogged tiles:

![image](https://github.com/wesnoth/wesnoth/assets/4893136/db9b47b8-55df-4109-8ce8-155b123c7fc5)

**Comment by gfgtdf** [comment]
Actually they were already there before, they are just more noticable when the other artifacts are gone.

**Comment by gfgtdf** [comment]
Also they _sometimes_ disappear when you move the mouse over the tile :s

**Comment by gfgtdf** [comment]
Here a better picture of the fog artifacts 

![image](https://github.com/wesnoth/wesnoth/assets/4893136/9d8001a7-d9fa-4ee6-bdf5-3e2015e886c5)

And here another picture after the map was scrolled upwards, you can see the artifacts only on the bottom area that was onscreen while the zoom resolution was changed and at the top row:

![image](https://github.com/wesnoth/wesnoth/assets/4893136/1b84fcde-2b94-4ee1-8f59-d35dadd0f81c)

**Comment by gfgtdf** [comment]
Also you can see a thick grey line on the left side of the screen that shouldnt be there.

**Comment by gfgtdf** [comment]
Sooo this really makes me wonder how our terrain drawing code currently works, there seems to be "layers" and "layer groups" so that when a map is drawn, it draws first the first layergroup for all locations and then draws the second  layergroups for all locations etc.
BUT:
 if i read the code correctly, this only happens when all those locations are invalidated a the same time, so that when only one location is invalidated only that location is drawn, which means this layergroup system actually makes no sense at all.

Am i missing something?

**Comment by gfgtdf** [comment]
> Am i missing something?

what i was missing is `display::propagate_invalidation`which makes sure that when one tile of a big unit like a dragon is invalidated, all tiles are.

**Comment by Pentarctagon** [comment]
Though it's perhaps not aggressive enough in triggering redraws in the "big unit" case, given #7626.

**Comment by mesilliac** [comment]
Assuming the fog is also using the +0.25px scale tweak, it might need the offset to the left as well. Scaling up could push everything to fill holes on the right, but open gaps on the left. Adding a small negative offset along with the scale should center it and fill gaps on both sides.

It could also be incorrectly using linear interpolation, which would mix with empty pixels on the sides and cause seams.

**Comment by gfgtdf** [comment]
> Assuming the fog is also using the +0.25px scale

It also happens a little without the changed scale, but with the scale is even more, I think the problem is that the fog draws "twice" in that area making it darker, that the scaled fog images overlap (not that there are holes)

**Comment by mesilliac** [comment]
> I think the problem is that the fog draws "twice" in that area making it darker, that the scaled fog images overlap (not that there are holes)

If that's the case, then the scale hack could be done the other way around - that is, reduce the size of the source region in stead of increasing the size of the target region. That would need a tiny offset to ensure the source region was still centered. It also needs some care to ensure the scaling is done in target pixel scale not source pixel scale.

**Comment by gfgtdf** [comment]
> If that's the case, then the scale hack could be done the other way around - that is, reduce the size of the source region in stead of increasing the size of the target region. That would need a tiny offset to ensure the source region was still centered. It also needs some care to ensure the scaling is done according to target pixel scale not source pixel scale.

I'm sorry i don't understand this at all.

**Comment by mesilliac** [comment]
It is annoyingly complex to understand properly. Honestly i normally just jiggle things around randomly until it works.

I made an image explaining this case. Case A, the blit target size is increased. Case B the source region for the blit is decreased. Both end up with the target hex sourcing from a smaller interior region of the source texture.

![blit_targets](https://github.com/wesnoth/wesnoth/assets/243646/cccf6cfd-04ab-42f0-a25d-018725ace693)

**Comment by gfgtdf** [comment]
So SDL_RenderCopyF doesn't seem to support floating point source rectangles, so to me it looks like this approach wouldn't work, and using more than a pixel will also not work since it means that there will be pixels on the source which will just not to be drawn on the target.

**Comment by gfgtdf** [comment]
> Honestly i normally just jiggle things around randomly until it works.

Also if I do this it might still happen on a different machine, especially if I don't know why it happens or how the scaling algorithm actually works.

**Comment by mesilliac** [comment]
I normally would definitely not recommend this method, but in this particular case the solution is stable in that if you tweak a variable in the wrong direction it always makes things worse, and if you get it working reliably for you it's very unlikely to break things for someone else.

It could be problematic that the source region can't be tweaked. But there might still be a case which solves the problem robustly using only the destination position and scale.

There are still other solutions (rendering at 1:1 then scaling, going back to prerendering everything scaled to each zoom level, filling in some of the transparent areas around hexes) but the best solution by far if it can be made to work is tweaking the draw call.

**Comment by mesilliac** [comment]
So i looked closely to see if i could find any seams at different zoom levels that might indicate a problem on my system as well, and i do at some zoom levels get the zig-zag overlaps in fog. I think that might actually be a separate issue.

While i was doing that i spotted a zoom level that is reporting the wrong scale. It looks like the 0.3x zoom level is scaling to 23 pixels in stead of 24. This could definitely be causing holes. 0.5x is correctly reporting 36 pixels however.

The zoom scale is reported using `--log-info=display`

**Comment by mesilliac** [comment]
and the reason for that is in `data/game_config.cfg` where it sets the zoom level as "0.33333333333333333". This is, of course, cast directly to an integer after multiplying by 72 in `game_config.cpp`, and no matter how many 3s you put on the end it will still floor to the wrong value. Changing this to "0.334" may improve the 1/3x zoom result.

**Comment by mesilliac** [comment]
```
	// hex_size() is always a multiple of 4
	// and hex_width() a multiple of 3,
	// so there shouldn't be off-by-one-errors
	// due to rounding.
	// To display a hex fully on screen,
	// a little bit extra space is needed.
	// Also added the border two times.
```
this comment is in `display.cpp` and has been there since 2007. Several `hex_size` and `hex_width` violate this.

It looks like the grey borders in fog are caused by `hex_width`, which is defined as `hex_size * 3 / 4`, flooring to something wrong when `hex_size` (= `zoom_scale` = the height of one hex in pixels) is not a multiple of 4. There are several zoom levels for which this is the case. In these cases hexes are drawn too close togther and do in fact overlap.

At a quick glance this applies to zoom levels 0.25, 0.75 and 1.25 which do all seem to have the fog border problem for me.

**Comment by Wedge009** [comment]
> While i was doing that i spotted a zoom level that is reporting the wrong scale. It looks like the 0.3x zoom level is scaling to 23 pixels in stead of 24.

0.3333 x 72 = 23 is the only tile size that is not calculated accurately based on the default zoom levels in `game_config.cfg`. The other zoom levels result in a correct calculation, ie no truncation from the cast to integer. Still worth fixing for that one case, though.

I tried changing the definition of `hex_width()` in `display.hpp` to `static_cast<int>(std::round(zoom_ * 3.0 / 4.0))` but I still see gaps at certain zoom levels. If anything, 1.25x (90 px) seems to have much more noticeable gaps. I suppose the point is that the alignment is very hard to get  right.

**Comment by mesilliac** [comment]
The fog does seem to be a separate issue. I tried changing the definition of zoom levels in `src/game_config.cpp:277` to  `return std::stoi(zoom);` in stead of `return static_cast<int>(std::stold(zoom) * tile_size);`, and setting zoom levels in `data/game_config.cfg` to `zoom_levels = 16, 24, 36, 52, 72, 100, 144, 216, 288`. This completely fixes the fog issue for me as now all hex sizes are cleanly divisible by 4.

I still get no black holes, even with the non-simple hex size multiples like 52 and 100.

**Comment by mesilliac** [comment]
I was able to get some results with holes using particularly offensive zoom levels like 71 pixels. But i'm not sure those should even be fixed. For the divisible-by-four zoom levels i can't find any that miss pixels on my setup.

**Comment by Wedge009** [comment]
Agreed that it's not necessary to do anything at non-standard sizes like 71px.

My current monitor is 'only' ~163ppi. Testing #8014 I wasn't expecting to see any difference but it looks like the only zoom level I now see obvious artefacts is at 36 (0.5x). I also did a side-by-side comparison with 1.16 just to verify what I was looking at.

**Comment by Wedge009** [comment]
I feel like #8014 does enough to mitigate this (at least on my hardware) that it's not a blocker for 1.18, although it is not resolved completely.

**Comment by mesilliac** [comment]
As much as i'd like to fix it properly, i'd need to do extensive testing on a machine which can reproduce the problem to determine what the precise nature of the edge case is.

Re the worry of breaking other things with a scale/offset tweak - the highest supported pixel scale multiplier is 4, so tweaks of less than 0.125 pixels should either cause no change on any system, or cause a 1 pixel shift on all systems.

With the fog fix, the original tweak above might even work. If it does work with say a 0.25px tweak, try with something tiny like 0.001px. It should resolve the ambiguity just as well, with less chance of changing anything for anyone else.

**Comment by Wedge009** [comment]
Given the age of this, as well as the unlikelihood of getting a reliable resolution that works for _everyone_ - and _reliably_ at that - I think it's fine to close this. I just had a quick look at all zoom levels on 1fc57a490d4 (1.19.10+dev) with fog enabled and didn't notice anything. I accept that others may have different experiences on their hardware so if this is still a bothersome issue for them, feel free to re-open.

**Comment by sevu** [comment]
Can reproduce the little black dots with zooming out twice. It's easiest to notice over sand. I could test any changes.

Not sure if it makes sense to keep this open though.
